src/controller/CommissioningWindowOpener.cpp

```diff
@@ -0,0 +1,348 @@
+    }
+}
+
+void CommissioningWindowOpener::OnOpenCommissioningWindowFailure(void * context, CHIP_ERROR error)
+{
+    ChipLogError(Controller, "Failed to open pairing window on the device. Status %" CHIP_ERROR_FORMAT, error.Format());
+    auto * self     = static_cast<CommissioningWindowOpener *>(context);
+    self->mNextStep = Step::kAcceptCommissioningStart;
+    if (self->mCommissioningWindowCallback != nullptr)
+    {
+        self->mCommissioningWindowCallback->mCall(self->mCommissioningWindowCallback->mContext, self->mNodeId, error,
+                                                  SetupPayload());
+    }
+    else if (self->mBasicCommissioningWindowCallback != nullptr)
+    {
+        self->mBasicCommissioningWindowCallback->mCall(self->mBasicCommissioningWindowCallback->mContext, self->mNodeId, error);
+    }
+}
+
+void CommissioningWindowOpener::OnDeviceConnectedCallback(void * context, OperationalDeviceProxy * device)
+{
+    auto * self = static_cast<CommissioningWindowOpener *>(context);
+
+#if CHIP_ERROR_LOGGING
+    const char * messageIfError = nullptr;
+#endif // CHIP_ERROR_LOGGING
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    switch (self->mNextStep)
+    {
+    case Step::kReadVID: {
+        constexpr EndpointId kBasicClusterEndpoint = 0;
+        BasicCluster cluster;
+        cluster.Associate(device, kBasicClusterEndpoint);
+        err = cluster.ReadAttribute<app::Clusters::Basic::Attributes::VendorID::TypeInfo>(context, OnVIDReadResponse,
+                                                                                          OnVIDPIDReadFailureResponse);
+#if CHIP_ERROR_LOGGING
+        messageIfError = "Could not read VID for opening commissioning window";
+#endif // CHIP_ERROR_LOGGING
+        break;
+    }
+    case Step::kReadPID: {
+        constexpr EndpointId kBasicClusterEndpoint = 0;
+        chip::Controller::BasicCluster cluster;
+        cluster.Associate(device, kBasicClusterEndpoint);
+        err = cluster.ReadAttribute<app::Clusters::Basic::Attributes::ProductID::TypeInfo>(context, OnPIDReadResponse,
+                                                                                           OnVIDPIDReadFailureResponse);
+#if CHIP_ERROR_LOGGING
+        messageIfError = "Could not read VID for opening commissioning window";
+#endif // CHIP_ERROR_LOGGING
+        break;
+    }
+    case Step::kOpenCommissioningWindow: {
+        err = self->OpenCommissioningWindowInternal(device);
+#if CHIP_ERROR_LOGGING
+        messageIfError = "Could not connect to open commissioning window";
+#endif // CHIP_ERROR_LOGGING
+        break;
+    }
+    case Step::kAcceptCommissioningStart: {
+        err = CHIP_ERROR_INCORRECT_STATE;
+#if CHIP_ERROR_LOGGING
+        messageIfError = "Just got a connected device; how can we be done?";
+#endif // CHIP_ERROR_LOGGING
+        break;
+    }
+    }
+
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(Controller, "%s", messageIfError);
```

Comment:
Should we include the error here since it contains a stack trace on larger devices? Like

```
ChipLogError(Controller, "Error %" CHIP_ERROR_FORMAT ": %s", err.Format(), messageIfError);
```

Reply:
Good idea, done.